### DISCUSSION TITLE
Advice on how to accept custom file types?

### DISCUSSION THREAD

**Discussion by sticknuder** [rr-/szurubooru#618] [Q&A] [answered]
Hi, I host a Szurubooru instance for an animation and art creation program community.

My site allows community members to post art and animation exported from the program, but I would like to also support posting assets made with the program.

Specifically I'm looking to accept 3 proprietary file types:
.nodes (stickfigure)
.stknds (project)
.nodemc (movieclip)

I may down the line want to write custom code to render a preview for these filetypes, but for now the behavior I think could work is the following:

1. Be allowed to upload .nodes, .stknds, and .nodemc files
2. .nodes will be given a post type of stickfigure, .stknds will be given a post type of project, and .nodemc will be given a post type of movieclip
3. Each custom file type will have it's own icon on the post listing page (like how videos have the video icon)
4. The set content feature can be used to set the preview of the custom file since there is no way to render it automatically currently 
5. Each custom file page will default to a fallback image if no content has been set yet
6. The actual file you get when you click the download button on the post page is the actual asset file instead of the preview

I have a very rough idea of how to start developing this functionality, but if anyone has any advice before I get into the weeds of this, it would be much appreciated.

**Comment by TiredSysOp** [comment]
I imagine a self hosted file sharing platform might be better, if you do want to fork szurubooru then I'm sure there are plenty of resources that would allow you to handle those file types.

If it's based on pivot then there should be open source players and renderers that can handle those file types. This is way beyond the scope of szurubooru and the Devs here probably won't be able to help other than possibly help point you in the right direction.

Would be less work and more efficient to petition to their (sticknode) developers to add gallery functionality to their site or hunt through their forums for like minded individuals. Otherwise search the szurubooru issues and code for how filetypes are handled and how you can pipe them to a renderer or encoder.

**Reply by sticknuder** [reply]
Thanks for taking the time to reply

> I imagine a self hosted file sharing platform might be better, if you do want to fork szurubooru then I'm sure there are plenty of resources that would allow you to handle those file types.

Maybe, but I already have a private fork of Szurubooru that is live, public, growing, and already has a handful of users, so I don't feel like switching now.

> If it's based on pivot then there should be open source players and renderers that can handle those file types.

Down the line I might worry about rendering the assets, but I'm not worried about that right now.

> This is way beyond the scope of szurubooru and the Devs here probably won't be able to help other than possibly help point you in the right direction.

I don't think it's way beyond the scope of Szurubooru. The advice I'm looking for is advice on how I should go about modifying Szurubooru's source code to get my desired behavior out of Szurubooru.

But I think I get what you're saying. The devs probably don't want to waste time thinking about my specific use case for Szurubooru.

> Would be less work and more efficient to petition to their (sticknode) developers to add gallery functionality to their site or hunt through their forums for like minded individuals. 

The official Stick Nodes site already supports asset uploads, but they are curated and filtered of any nsfw assets. I want to allow nsfw assets.

> Otherwise search the szurubooru issues and code for how filetypes are handled and how you can pipe them to a renderer or encoder.

Closest I've found so far from my searching is PR #593, which I thought may allow for the configuration of allowed filetypes in the sense of being able to add more, but it unfortunately seems to only allow you to restrict which filetypes are accepted.

**Comment by sticknuder** [comment, marked as answer]
I traced the flow of uploads from the uploads tpl to where filetypes are actually handled and extended it to handle .nodes, .nodemc, and .stknds. My implementation is a little hacky, but it works, so I'm satisfied :)
